### ISSUE TITLE
Docker image

### ISSUE THREAD

**Issue by Levino** [Daplie/node-letsencrypt#12]
Are there plans to docker this? I will integrate this service into my docker image as soon as possible, just wanted to avoid maintaining two repos, if you guys also wanted to build a docker image. https://github.com/Levino/letsencrypt-cert-service

**Comment by Levino** [comment]
Reading up more I do not think that you guys support "certificate" download and so forth, right? Then this package does not provide all I need but might still be a good starting point for my node service in the docker image.

**Comment by coolaj86** [comment]
We do support certificate download. It goes into the directory specified by `--config-dir` (letsencrypt-cli) or `configDir` (node-letsencrypt).

We do want to get this on docker. This whole week we're busy at CES, but I'd love to work with you on it after that.

**Comment by coolaj86** [comment]
Feel free to shoot me some specific questions and when I have some downtime at the booth I'll write code examples.

**Comment by coolaj86** [comment]
Also, just read your README, this tool and the official client both have the webroot option, so haproxy is no longer necessary in most situations.

**Comment by Levino** [comment]
How would you use webroot with a microservices setup where all traffic gets handled by many microservices which do not even have persistent storage (except for database access) and to NOT serve static http files?

**Comment by coolaj86** [comment]
Yes... I'm at CES this week and pretty busy, but there is some abstraction in place for that to be possible. Take a look at [letiny-core](https://github.com/Daplie/letiny-core).

**Comment by coolaj86** [comment]
Hey @Levino, are you still interested in working on this? I have some free cycles later this week and would be interested in collaborating.

**Comment by Levino** [comment]
@coolaj86 Yes, I am indeed. Please take a look at my attempt: https://github.com/Levino/letsencrypt-cert-service

There are a couple of issues:

- I am using the letsencrypt cli to generate the certificates. I would prefer a node only solution
- Delivering the ssl certificates and keys from behind a load balancer is tricky. The easy solution would be to have two servers running in the container, one delivering unencrypted traffic for the communication with letsencrypt and one delivering encrypted traffic for the admin of the domain. Unfortunately with haproxy / tutum load balancing one can only load balance to one port in a container. I need to find an elegant solution to this.

I am happy to help when I can, but I don't have much spare time this or next week.

**Comment by Seikho** [comment]
`letsencrypt-cli` is a node only solution.
This particular project isn't suitable for a Docker container, but `letsencrypt-express` or `letsencrypt-cli` are.

**Comment by coolaj86** [comment]
@Levino see https://github.com/Daplie/node-letsencrypt/issues/39

Now that there is a storage plugin architecture (and one for challenge plugins as well) you can do exactly what you were suggesting. See the new README (v2).

**Comment by kathiedart** [comment]
Was there any progress made here since?I'm interested too and can do something similar.

**Comment by Levino** [comment]
I recommend docker swarm mit traefik for ssl terminaton and load balancing. Letsencrypt is baked in and works like a charm!

**Comment by coolaj86** [comment]
If anyone wants to help with this, please get in contact with me over at the official repo https://git.coolaj86.com/coolaj86/greenlock.js
